build/patches/add-user-scripts.v5.patch

```diff
@@ -0,0 +1,9810 @@
++        script->set_name_space(value);
++      } else if (GetDeclarationValue(line, kNameDeclaration, &value)) {
++        script->set_name(value);
++      } else if (GetDeclarationValue(line, kVersionDeclaration, &value)) {
++        base::Version version(value);
++        if (version.IsValid())
++          script->set_version(version.GetString());
++      } else if (GetDeclarationValue(line, kDescriptionDeclaration, &value)) {
++        script->set_description(value);
++      } else if (GetDeclarationValue(line, kMatchDeclaration, &value)) {
++        URLPattern pattern(UserScript::ValidUserScriptSchemes());
++        if (URLPattern::ParseResult::kSuccess != pattern.Parse(value)) {
++          error_message = "Invalid UserScript Schema " + value;
++          return false;
++        }
++        script->add_url_pattern(pattern);
++      } else if (GetDeclarationValue(line, kExcludeMatchDeclaration, &value)) {
++        URLPattern exclude(UserScript::ValidUserScriptSchemes());
++        if (URLPattern::ParseResult::kSuccess != exclude.Parse(value)) {
++          error_message = "Invalid UserScript Schema " + value;
++          return false;
++        }
++        script->add_exclude_url_pattern(exclude);
++      } else if (GetDeclarationValue(line, kRunAtDeclaration, &value)) {
++        if (value == kRunAtDocumentStartValue)
++          script->set_run_location(UserScript::DOCUMENT_START);
++        else if (value == kRunAtDocumentEndValue)
++          script->set_run_location(UserScript::DOCUMENT_END);
++        else if (value == kRunAtDocumentIdleValue)
++          script->set_run_location(UserScript::DOCUMENT_IDLE);
++        else {
++          error_message = "Invalid RunAtDeclaration " + value;
++          return false;
++        }
++      }
++
++      // TODO(aa): Handle more types of metadata.
++    }
++
++    line_start = line_end + 1;
++  }
++
++  // If no patterns were specified, default to @include *. This is what
++  // Greasemonkey does.
++  if (script->globs().empty() && script->url_patterns().is_empty())
++    script->add_glob("*");
++
++  return true;
++}
++
++// static
++bool LoadUserScriptFromFile(
++    const base::FilePath& user_script_path, const GURL& original_url,
++    std::unique_ptr<UserScript>& script,
++    base::string16* error) {
++
++    std::string script_key = user_script_path.BaseName().value();
++
++    std::string content;
++    if (user_script_path.IsContentUri()) {
++      //LOG(INFO) << "---Path " << user_script_path << " is a content uri";
++
++      base::FilePath tempFilePath;
++      if( base::CreateTemporaryFile(&tempFilePath) == false ) {
++        *error = base::ASCIIToUTF16("Could not read create temp file.");
++        return false;
++      }
++
++      if( base::CopyFile(user_script_path, tempFilePath) == false ) {
++        *error = base::ASCIIToUTF16("Could not copy file to temp file.");
++        return false;
++      }
++
++      if (!base::ReadFileToString(tempFilePath, &content)) {
```

Comment:
if you see here https://source.chromium.org/chromium/chromium/src/+/master:base/files/file_util_posix.cc;l=1124
in case it is a ContentUri, the code runs on java and does too many things, copying it in a temporary one simplifies my life